app/code/Magento/Checkout/Test/Unit/Plugin/ResetQuoteAddressesTest.php

```diff
@@ -0,0 +1,172 @@
+            'getAllAddresses',
+            'getAllVisibleItems',
+            'removeAddress',
+            'getExtensionAttributes',
+            'isVirtual',
+        ]);
+        $this->extensionAttributesMock = $this->getMockBuilder(CartExtensionInterface::class)
+            ->setMethods(
+                [
+                    'getShippingAssignments',
+                    'setShippingAssignments'
+                ]
+            )
+            ->getMockForAbstractClass();
+
+        $this->plugin = new ResetQuoteAddresses();
+    }
+
+    /**
+     * Test removing the addresses from a non empty quote
+     */
+    public function testRemovingTheAddressesFromNonEmptyQuote()
+    {
+        $quoteVisibleItems = [1, 2];
+
+        $this->quoteMock->expects($this->any())
+            ->method('getAllVisibleItems')
+            ->will($this->returnValue($quoteVisibleItems));
+        $this->quoteMock->expects($this->never())
+            ->method('getAllAddresses')
+            ->willReturnSelf();
+
+        $this->plugin->afterRemoveItem($this->quoteMock, $this->quoteMock, 1);
+    }
+
+    /**
+     * Test clearing the addresses from an empty quote
+     *
+     * @dataProvider quoteDataProvider
+     * @param bool $isVirtualQuote
+     * @param bool $quoteHasAddresses
+     * @param $extensionAttributes
+     */
+    public function testClearingTheAddressesFromEmptyQuote(
+        bool $isVirtualQuote,
+        bool $quoteHasAddresses,
+        $extensionAttributes
+    ) {
+        $quoteVisibleItems = [];
+
+        $this->quoteMock->expects($this->any())
+            ->method('getAllVisibleItems')
+            ->will($this->returnValue($quoteVisibleItems));
+
+        if ($quoteHasAddresses) {
+            $address = $this->createPartialMock(Address::class, ['getId']);
+
+            $address->expects($this->any())
+                ->method('getId')
+                ->willReturn(1);
+
+            $addresses = [$address];
+
+            $this->quoteMock->expects($this->any())
+                ->method('getAllAddresses')
+                ->will($this->returnValue($addresses));
+
+            $this->quoteMock->expects($this->exactly(count($addresses)))
+                ->method('removeAddress')
+                ->willReturnSelf();
+        } else {
+            $this->quoteMock->expects($this->any())
+                ->method('getAllAddresses')
+                ->willReturn([]);
+        }
+
+        $this->quoteMock->expects($this->once())
+            ->method('getExtensionAttributes')
+            ->willReturn($this->extensionAttributesMock);
+
+        $this->quoteMock->expects($this->once())
+            ->method('isVirtual')
+            ->willReturn($isVirtualQuote);
+
+        if ($isVirtualQuote && $extensionAttributes) {
+            $this->extensionAttributesMock->expects($this->any())
+                ->method('getShippingAssignments')
+                ->willReturn([1]);
```

Comment:
If that `1` means the same ID that `STUB_ADDRESS_ID` - put that reference there instead of magic number that has unclear meaning.

Reply:
Actually no, so I moved it to a const as well.